modules, homestakeros: autogen ssv keys and then some

tupakkatapa opened the pull request:
Proposal, not tested. Aiming to reduce the threshold for setting up a SSV. 

Keys would be generated automatically at boot if they do not exist, and `addons.ssv-node.dataDir` is a persistent location. The `dataDir` option defaults to the first mount option's "where" path + /ssv subdirectory; if none of the mounts exist, it falls back to /mnt/addons/ssv, however this fallback path would not result in automatic key generation since this location is probably in RAM.

Lastly, the node would only activate when the operator is properly registered. Registration is checked by querying the SSV Network API in 10 minute intervals.

tupakkatapa commented:
@jhvst thoughts?

jhvst commented on nixosModules/homestakeros/options.nix:
```diff
@@ -1,6 +1,19 @@
-}: {
+}:
+let
+  # Try to get the first mount's path, fallback to null
```
Why default to null when you match on it later on to make it a string anyway?
Null is an anti-pattern: your return "type" is basically `Either` ( `String` | *Undefined* ).
Null is very bad because it has so much undefined behavior when its being ran against library functions, especially in languages like Nix which do not have strong typing.

jhvst commented on nixosModules/homestakeros/options.nix:
```diff
@@ -1,6 +1,19 @@
+  firstMountPath =
+    let
+      mountNames = lib.attrNames (lib.filterAttrs (name: mount: mount.enable) cfg.mounts);
+      firstMountName = lib.optional (mountNames != [ ]) (builtins.head mountNames);
```
Make sure `mountNames` always returns an array.

jhvst commented on nixosModules/homestakeros/options.nix:
```diff
@@ -1,6 +1,19 @@
+      mountNames = lib.attrNames (lib.filterAttrs (name: mount: mount.enable) cfg.mounts);
+      firstMountName = lib.optional (mountNames != [ ]) (builtins.head mountNames);
+    in
+    if firstMountName != [ ]
+    then "${cfg.mounts.${builtins.head firstMountName}.where}"
+    else null;
```
Because when `mountNames` is always an array, you can always call array functions like head on it.

jhvst commented on nixosModules/homestakeros/default.nix:
```diff
@@ -237,26 +233,54 @@ in
+                  # Check if keys exist
+                  if [ ! -f "${cfg.addons.ssv-node.privateKeyFile}" ] || [ ! -f "${cfg.addons.ssv-node.publicKeyFile}" ]; then
+
+                    # Check if dataDir is on tmpfs
+                    FS_TYPE=$(df --output=fstype "${cfg.addons.ssv-node.dataDir}" | tail -n1)
+                    if [[ "$FS_TYPE" =~ ^(tmpfs|overlay|squashfs)$ ]]; then
+                      echo "error: '${cfg.addons.ssv-node.dataDir}' is on a tmpfs; generated keys would be lost after reboot"
+                      exit 1
+                    fi
```
Notice that you could have just use the `cfg` to track down the `dataDir` to the corresponding `systemd-mount` option hence get static guarantees rather than do this runtime checking.

jhvst commented on packages/init-ssv/init-ssv.sh:
lgtm

jhvst commented on docs/homestakeros/3.2-ssv_node.md:
````diff
@@ -13,30 +13,29 @@ Running an SSV node allows you to participate in distributed validator technolog
-2. **Generate a key pair**:
+2. **Retrieve your public key**:
 
    ```bash
-   nix run github:ponkila/HomestakerOS#init-ssv --accept-flake-config -- <MY_OPERATOR_PASSWORD>
+   cat /path/to/ssv_operator_key.pub
    ```
 
-   This command will generate three files: a `password` file containing your password, an encrypted private key in `ssv_operator_key`, and a public key in `ssv_operator_key.pub`. You'll need the public key for registration.
-
-   The key pair is essential for the secure operation of your SSV node. The private key must be kept secure and will be used to sign messages, while the public key identifies your operator on the network.
+   This public key is essential for the operator registration process in the next section. Make sure to copy it exactly as shown.
````
Retrieve it on which host? Could they instead run some Nix command?

jhvst commented on docs/homestakeros/3.2-ssv_node.md:
````diff
@@ -58,63 +57,43 @@ You can do this directly through the HomestakerOS Web UI:
+2. **Check the service status**:
 
    ```bash
    systemctl status ssv-autostart.service
````
Notice you changed the name of the unit file later in the PR.

tupakkatapa commented:
Since we need to be able to evaluate schema, circular import of cfg is unintuitive and results in an error. I suggest simplifying the approach by just removing the key file options and dataDir's first mount default completely.

```bash
[I] kari@torgue ~/W/p/HomestakerOS (jesse/improve-init-ssv)> nix eval --json .#schema | jq -r '.'
error:
       … while evaluating attribute 'addons'

       … while evaluating attribute 'ssv-node'

       … while evaluating attribute 'dataDir'

       … while evaluating attribute 'default'
         at /nix/store/50jy2msjzq74j5dngf8kld5r2bvmdb0f-source/introspect.nix:89:11:
           88|           type = typeString;
           89|           default = v.default or null;
             |           ^
           90|           description = v.description or null;

       … while evaluating the module argument `cfg' in "/nix/store/50jy2msjzq74j5dngf8kld5r2bvmdb0f-source/nixosModules/homestakeros/options.nix":

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: attribute 'cfg' missing
       at /nix/store/096c6m1vzz4zr46lhqc8c1ynwr4zpc0a-source/lib/modules.nix:515:28:
          514|         addErrorContext (context name)
          515|           (args.${name} or config._module.args.${name})
             |                            ^
          516|       ) (functionArgs f);
```

tupakkatapa commented:
Or then just scrap this for now and do this properly when agenix-rekey gets implemented.

jhvst commented:
> Since we need to be able to evaluate schema, circular import of cfg is unintuitive and results in an error. I suggest simplifying the approach by just removing the key file options and dataDir's first mount default completely.
> 
> ```shell
> [I] kari@torgue ~/W/p/HomestakerOS (jesse/improve-init-ssv)> nix eval --json .#schema | jq -r '.'
> error:
>        … while evaluating attribute 'addons'
> 
>        … while evaluating attribute 'ssv-node'
> 
>        … while evaluating attribute 'dataDir'
> 
>        … while evaluating attribute 'default'
>          at /nix/store/50jy2msjzq74j5dngf8kld5r2bvmdb0f-source/introspect.nix:89:11:
>            88|           type = typeString;
>            89|           default = v.default or null;
>              |           ^
>            90|           description = v.description or null;
> 
>        … while evaluating the module argument `cfg' in "/nix/store/50jy2msjzq74j5dngf8kld5r2bvmdb0f-source/nixosModules/homestakeros/options.nix":
> 
>        (stack trace truncated; use '--show-trace' to show the full, detailed trace)
> 
>        error: attribute 'cfg' missing
>        at /nix/store/096c6m1vzz4zr46lhqc8c1ynwr4zpc0a-source/lib/modules.nix:515:28:
>           514|         addErrorContext (context name)
>           515|           (args.${name} or config._module.args.${name})
>              |                            ^
>           516|       ) (functionArgs f);
> ```

Forgot about this, lets proceed with what we have.

jhvst reviewed (approved):
lgtm

tupakkatapa commented:
Tested persistent and registeration check, fixed a syntax error. Set:
- `addons.ssv-node.dataDir` -> `/mnt/ethereum/addons/ssv`
- `mounts.<name>.where` -> `/mnt/ethereum`

```bash
core@homestakeros-test ~> ls /mnt/ethereum/addons/ssv
password  ssv_operator_key  ssv_operator_key.pub
core@homestakeros-test ~> journalctl -fu ssv-node.service
Apr 07 18:01:27 homestakeros-test systemd[1]: Started Operator node for Secret Shared Validators (SSV).
Apr 07 18:01:27 homestakeros-test ssv-node-start[974]: error: operator is not registered yet, exiting
Apr 07 18:01:27 homestakeros-test systemd[1]: ssv-node.service: Main process exited, code=exited, status=1/FAILURE
Apr 07 18:01:27 homestakeros-test systemd[1]: ssv-node.service: Failed with result 'exit-code'.
Apr 07 18:01:27 homestakeros-test systemd[1]: ssv-node.service: Consumed 29ms CPU time, 5.3M memory peak, 2.2M read from disk.
Apr 07 18:11:27 homestakeros-test systemd[1]: ssv-node.service: Scheduled restart job, restart counter is at 1.
Apr 07 18:11:27 homestakeros-test systemd[1]: Started Operator node for Secret Shared Validators (SSV).
Apr 07 18:11:28 homestakeros-test ssv-node-start[1479]: error: operator is not registered yet, exiting
Apr 07 18:11:28 homestakeros-test systemd[1]: ssv-node.service: Main process exited, code=exited, status=1/FAILURE
Apr 07 18:11:28 homestakeros-test systemd[1]: ssv-node.service: Failed with result 'exit-code'.
Apr 07 18:11:28 homestakeros-test systemd[1]: ssv-node.service: Consumed 84ms CPU time, 5.5M memory peak, 11K read from disk, 4.1K incoming IP traffic, 1.9K outgoing IP traffic.
```

Switched to registered key at this point..

```bash
Apr 07 18:21:28 homestakeros-test systemd[1]: ssv-node.service: Scheduled restart job, restart counter is at 2.
Apr 07 18:21:28 homestakeros-test systemd[1]: Started Operator node for Secret Shared Validators (SSV).
Apr 07 18:21:29 homestakeros-test ssv-node-start[2073]: operator is registered, starting ssv node...
Apr 07 18:21:29 homestakeros-test ssv-node-start[2079]: 2025-04-07T15:21:29.938032Z        INFO        starting SSV-Node:v2.0.0-e39f1d0ec67c47f6ba437d83367aba0d2b4d34dd
...
```

Set:
- `addons.ssv-node.dataDir` -> `/mnt/addons2/ssv`
- `mounts.<name>.where` -> `/mnt/ethereum`

```bash
core@homestakeros-test ~> ls /mnt/addons2/ssv
ls: cannot access '/mnt/addons2/ssv': No such file or directory
core@homestakeros-test ~ [2]> systemctl status ssv-node.service
Unit ssv-node.service could not be found.
```
